File: android/app/src/main/kotlin/org/gaindrive/android/data/AccountLimits.kt
package org.gaindrive.android.data

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withTimeoutOrNull
import org.gaindrive.android.data.model.ServerConfig
import org.gaindrive.android.data.model.ServerId
import org.gaindrive.android.net.SubsonicClientFactory
import org.gaindrive.android.net.requireOk
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Each server's `maxBitRate` ceiling for the signed-in account.
 *
 * The server enforces this whatever the client asks for, so the app has to know
 * it before it can name the quality it is about to receive — a request for the
 * original file on a capped account comes back as MP3 at the cap, and bytes
 * stored under the wrong quality key are worse than bytes not stored at all.
 *
 * Held in memory for the process rather than persisted: it is one cheap call
 * per server per launch, it picks up a change made on the server without any
 * invalidation logic, and there is nothing to migrate. The first stream URL of
 * a session waits for it, which is why the timeout is short — an unreachable
 * server means uncapped, and offline playback comes from the cache anyway.
 */
@Singleton
class AccountLimits @Inject constructor(
	private val clients: SubsonicClientFactory,
) {

	private val mutex = Mutex()
	private val caps = mutableMapOf<ServerId, Int>()

	/** 0 means no limit, and is also what every failure resolves to. */
	suspend fun capFor(config: ServerConfig): Int = mutex.withLock {
		caps.getOrPut(config.id) { fetch(config) }
	}

	private suspend fun fetch(config: ServerConfig): Int {
		val cap = withTimeoutOrNull(FETCH_TIMEOUT_MS) {
			try {
				clients.clientFor(config)
					.getUser(config.username)
					.requireOk()
					.user
					?.maxBitRate
			} catch (e: CancellationException) {
				throw e
			} catch (e: Exception) {
				// Uncapped is the right guess when we cannot ask: it is what
				// the overwhelming majority of accounts are, and guessing a cap
				// that is not there would degrade every stream on this server
				// for the rest of the session.
				null
			}
		}
		return cap?.takeIf { it > 0 } ?: 0
	}

	private companion object {
		const val FETCH_TIMEOUT_MS = 2_000L
	}
}

File: ios/GainDrive/Data/AccountLimits.swift
//	GainDrive for iOS
//	Copyright (C) 2026 Kasper Peeters
//
//	This file is part of GainDrive, distributed under the GNU General
//	Public License version 3 or later, with the Application Distribution
//	Exception in ios/LICENSE. See LICENSE at the repository root for the
//	full text of the GPL.

import Foundation

/// Each server's `maxBitRate` ceiling, asked once per process.
///
/// Held in memory rather than persisted: it is a property of the account as the
/// server currently has it, and a stale ceiling read from disk would cap every
/// stream for a session against a limit that may have been lifted.
///
/// **Uncapped is the right guess when we cannot ask.** Every failure — an
/// unreachable server, a refusal, the timeout — resolves to 0. Guessing a cap
/// that is not there would degrade every stream on that server for the rest of
/// the session, whereas guessing wrong in this direction costs nothing: the
/// server applies its own ceiling regardless of what the client asked for. The
/// only consequence is a cache key naming a quality the bytes are not, which is
/// a phase 5 concern and is why this is worth getting right now.
actor AccountLimits {
	/// The in-flight *task*, not the answer. An actor releases isolation at
	/// every `await`, so two tracks starting together would otherwise both find
	/// the cache empty and both ask.
	private var inFlight: [ServerId: Task<Int, Never>] = [:]

	private static let timeout = Duration.seconds(2)

	func cap(for server: ServerId, using clients: ServerClients) async -> Int {
		if let existing = inFlight[server] { return await existing.value }
		guard let client = clients.client(for: server) else { return 0 }

		let task = Task<Int, Never> { await Self.fetch(client) }
		inFlight[server] = task
		return await task.value
	}

	/// The first play of a session waits for this, which is why the timeout is
	/// short — and why `PlayerConnection` publishes `loadingRef` before
	/// reaching it.
	private static func fetch(_ client: SubsonicClient) async -> Int {
		await withTaskGroup(of: Int.self) { group in
			group.addTask {
				guard let user = try? await client.user(), let cap = user.maxBitRate, cap > 0
				else { return 0 }
				return cap
			}
			group.addTask {
				try? await Task.sleep(for: Self.timeout)
				// The same answer a failure gives, deliberately: a server that
				// did not answer in time is treated as having no ceiling.
				return 0
			}
			let first = await group.next() ?? 0
			group.cancelAll()
			return first
		}
	}
}
